src/lib.rs

```diff
@@ -0,0 +1,109 @@
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+use std::u32;
+
+mod convert;
+mod fmt;
+mod ops;
+
+/// An index into a string.
+///
+/// The index is stored as a 32 bit integer,
+/// assuming we only deal with text shorter than 4 GiB.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Copy, Clone, Default, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct StrIndex {
+    raw: u32,
+}
+
+impl StrIndex {
+    /// Index equal to the string length of this `char`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use str_index::*;
+    /// assert_eq!(
+    ///     StrIndex::from_char_len('😂'),
+    ///     StrIndex::from(4),
+    /// );
+    /// ```
+    pub fn from_char_len(c: char) -> Self {
+        StrIndex::from(c.len_utf8() as u32)
+    }
+
+    /// Index equal to the length of this string.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use str_index::*;
+    /// assert_eq!(
+    ///     StrIndex::from_str_len("メカジキ"),
+    ///     StrIndex::from(12),
+    /// );
+    /// ```
+    pub fn from_str_len(s: &str) -> Self {
+        assert!(s.len() < u32::MAX as usize, "string index too large");
+        StrIndex {
+            raw: s.len() as u32,
+        }
+    }
+
+    /// This index as a raw `usize`.
+    pub fn to_usize(self) -> usize {
+        self.into()
+    }
+
+    /// Checked integer addition.
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        self.raw.checked_add(rhs.raw).map(StrIndex::from)
+    }
+
+    /// Checked integer subtraction.
+    pub fn checked_sub(self, rhs: Self) -> Option<Self> {
+        self.raw.checked_sub(rhs.raw).map(StrIndex::from)
+    }
+}
+
+/// A range of a string, represented as a half-open range of `StrIndex`.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Copy, Clone, Default, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub struct StrRange {
+    start: StrIndex,
+    end: StrIndex,
+}
+
+impl StrRange {
+    /// The half-open range (`start..end`) between two points in a string.
+    pub fn between(start: StrIndex, end: StrIndex) -> Self {
+        let range = StrRange { start, end };
+        assert!(start <= end, "invalid string range {}", range);
+        range
+    }
+
+    /// The half-open range (`start..start+len`) between two points in a string.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use str_index::*;
+    /// assert_eq!(
+    ///     StrRange::from_len(10.into(), 10.into()),
+    ///     StrRange::between(10.into(), 20.into()),
+    /// );
+    /// ```
+    pub fn from_len(start: StrIndex, len: StrIndex) -> Self {
```

Comment:
That”s a name i want to change in text unit. Originally it was created for the symmetry with from_to, but, while from_to reads unambiguously, from_len can be read in two ways (as “from, len” and as “from this len”)

Reply:
Would it make sense to spell this like `StrRange::from(index).with_len(len)`? That's the only other reasonable spelling I can come up with.

Reply:
I've decided to remove this function from this PR and add it in a future PR.